daft/io/_parquet.py

```diff
@@ -68,10 +68,11 @@ def read_parquet(
             "Specifying schema_hints is deprecated from Daft version >= 0.3.0! Instead, please use the 'schema' and 'infer_schema' arguments."
         )
 
-    is_ray_runner = context.get_context().is_ray_runner
     # If running on Ray, we want to limit the amount of concurrency and requests being made.
     # This is because each Ray worker process receives its own pool of thread workers and connections
-    multithreaded_io = not is_ray_runner if _multithreaded_io is None else _multithreaded_io
+    multithreaded_io = (
+        (context.get_context().get_or_create_runner().name != "ray") if _multithreaded_io is None else _multithreaded_io
+    )
```

Comment:
So doing `set_runner` after dataframe operations like `read_parquet` that do `get_or_create_runner` can error, i.e.
```
import daft
df = daft.read_parquet("ordered_by_orderkey.parquet")
daft.context.set_runner_ray()   
df.collect()
```
This feels pretty dangerous and i can see users who may do this, i.e. they write the query first then at the end they set whatever runner and the collect.

I think for safety can we instead use `_get_runner_config_from_env` to check for runner in dataframe ops / expressions. The con here is that runner may change and maybe performance gets whacked a bit cuz we do multithreaded io on ray runner. (or some fixed shape tensor bug happens i guess ...)

Reply:
I actually think we should move to making this bad user behavior, and instead encourage users to call `daft.context.set_runner_*` only before any API calls to Daft.

Too much magic, make explicit is good